src/process.rs

```diff
@@ -637,15 +636,333 @@ mod unix {
             user.name
         }
     }
-}
 
+    //-------- EnvSockets ----------------------------------------------------
+
+    /// Accces to pre-bound sockets passed via environment variables.
+    pub struct EnvSockets {
+        /// An ordered collection of socket file descriptors along with their
+        /// address and type.
+        ///
+        /// Preserves the order the sockets were provided to us.
+        fds: Vec<SocketInfo>,
+    }
+
+    const SD_LISTEN_FDS_START: RawFd = 3;
+    const MAX_LISTEN_FDS: usize = 32; // Arbitrary limit against bad input.
+
+    impl EnvSockets {
+        /// Retrieve pre-bound socket identifiers from environment variables.
+        ///
+        /// Uses the following environment variables per [`sd_listen_fds()``]:
+        ///   - LISTEN_PID: Must match our own PID.
+        ///   - LISTEN_FDS: The number of FDs being passed to the application.
+        ///
+        /// The remaining FDs are numbered SD_LISTEN_FDS_START + n where
+        /// SD_LISTEN_FDS_START is defined as 3 in <systemd/sd-daemon.h>.
+        ///
+        /// If the `unset_environment` flag is true and the LISTEN_PID
+        /// variable matches our PID, the LISTEN_PID and LISTEN_FDS
+        /// environment variables will be removed before this function
+        /// returns.
+        ///
+        /// SystemD documentation makes no mention of a maximum value for
+        /// LISTEN_FDS. This function will limit the LISTEN_FDS value to 32
+        /// to prevent atempts to allocate large amounts of memory due to an
+        /// accidental or malicious large LISTEN_FDS value.
+        ///
+        /// Only sockets of type AF_INET UDP and AF_INET TCP, whose address can
+        /// be determined, will be captured by this function. Other socket file
+        /// descriptors will be ignored.
+        ///
+        /// [`sd_listen_fds()`]: https://www.man7.org/linux/man-pages/man3/sd_listen_fds.3.html#NOTES
+        ///
+        /// Safety:
+        /// =======
+        ///
+        /// This function is only safe to call in a single threaded context
+        /// as it calls [`std::env::remove_var()`].
+        pub fn from_env(unset_environment: bool) -> Self {
```

Comment:
Maybe it would be better to have a separate `clear_env` function so that you can call this one at any time?

This would also allow the function to return early and make everything easier …